[feature]: BumpFee: return raw tx hex on success

mohamedawnallah opened the pull request:
## Change Description

Fixes #8470 

## Demo

https://github.com/lightningnetwork/lnd/assets/69568555/c924a3c6-ea3c-44ac-ad64-cb22a0207355


## Steps to Test
Steps for reviewers to follow to test the change.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

guggero commented:
This is the wrong transaction being returned. What the PR currently returns is the transaction that should be bumped. But instead we'll want to return the new transaction the sweeper creates. For that you have to use the result channel returned from `UpdateParams` and wait for the result being sent. That could take a whole while though, depending on the user-configured sweeper interval.

mohamedawnallah commented:
Thanks @guggero for your mention and sorry for that I assumed that the transaction was mutated in place but after reading a bumpfee command it follows the `Replace by Fee Policy`.

guggero commented on lnrpc/rpc_utils.go:
```diff
@@ -234,3 +236,17 @@ func CalculateFeeRate(satPerByte, satPerVByte uint64, targetConf uint32,
+// SerializeAndEncodeTx serializes the given wire transaction and returns
+// its hexadecimal representation. If serialization fails, an error is
+// returned.
+func SerializeAndEncodeTx(tx wire.MsgTx) (string, error) {
```
nit: should say `SerializeAndHexEncodeTx` instead.

guggero commented on lnrpc/walletrpc/walletkit.proto:
```diff
@@ -1166,11 +1166,19 @@ message BumpFeeRequest {
     */
     uint64 sat_per_vbyte = 5;
+
+    // Indicates whether to include the raw transaction hex for
```
Need to mention that this will wait for the batcher to create a transaction, which might take a while, at least the duration of the configured `sweeper.batchwindowduration` value (default is 30 seconds).

guggero commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -985,11 +985,25 @@ func (w *WalletKit) BumpFee(ctx context.Context,
+		var sweepTxHex string
+		if in.IncludeRawTx {
+			select {
+			case sweepResult := <-sweepChan:
```
To make sure the user can cancel this operation (and to also fix the linter issue), we must listen on `ctx.Done()` here which fires when the user aborts on the client side (Ctrl+C). Same below.

guggero commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -227,6 +227,11 @@ var bumpFeeCommand = cli.Command{
 		},
+		cli.StringFlag{
+			Name: "include_raw_tx",
+			Usage: "include the raw transaction hex for " +
```
Same here re telling the user this takes a while.

guggero commented on itest/lnd_onchain_test.go:
```diff
@@ -278,9 +279,19 @@ func runCPFP(ht *lntest.HarnessTest, alice, bob *node.HarnessNode) {
-	bob.RPC.BumpFee(bumpFeeReq)
+	bumpFeeResp := bob.RPC.BumpFee(bumpFeeReq)
+
+	// The sweep  tx hex on success should be set and be valid.
```
nit: double space.

guggero reviewed (commented):
A few additional nits, but getting close.

mohamedawnallah commented:
@guggero Thanks a lot for the review. I've addressed your comments but one thing related to the `cpfp` on-chain test case I don't know why it fails.:
```
test case failed:
=== RUN   TestLightningNetworkDaemon/tranche00/01-of-1/btcd/cpfp
=== NAME  TestLightningNetworkDaemon
    harness_rpc.go:97:
        	Error Trace:	/Users/mohamedawnallah/Desktop/lnd/lntest/rpc/harness_rpc.go:97
        	            				/Users/mohamedawnallah/Desktop/lnd/lntest/rpc/wallet_kit.go:242
        	            				/Users/mohamedawnallah/Desktop/lnd/itest/lnd_onchain_test.go:285
        	            				/Users/mohamedawnallah/Desktop/lnd/itest/lnd_onchain_test.go:219
        	            				/Users/mohamedawnallah/Desktop/lnd/lntest/harness.go:353
        	            				/Users/mohamedawnallah/Desktop/lnd/itest/lnd_test.go:139
        	Error:      	Received unexpected error:
        	            	rpc error: code = DeadlineExceeded desc = context deadline exceeded
        	Messages:   	Bob: failed to call BumpFee
=== NAME  TestLightningNetworkDaemon/tranche00/01-of-1/btcd/cpfp
    harness.go:406: finished test: cpfp, start height=440, end height=441, mined blocks=1
    harness.go:435:
        	Error Trace:	/Users/mohamedawnallah/Desktop/lnd/lntest/harness.go:435
        	            				/opt/homebrew/Cellar/go/1.21.7/libexec/src/testing/testing.go:1169
        	            				/opt/homebrew/Cellar/go/1.21.7/libexec/src/testing/testing.go:1347
        	            				/opt/homebrew/Cellar/go/1.21.7/libexec/src/testing/testing.go:1589
        	            				/opt/homebrew/Cellar/go/1.21.7/libexec/src/runtime/panic.go:523
        	            				/opt/homebrew/Cellar/go/1.21.7/libexec/src/testing/testing.go:999
        	            				/Users/mohamedawnallah/Desktop/lnd/lntest/rpc/harness_rpc.go:97
        	            				/Users/mohamedawnallah/Desktop/lnd/lntest/rpc/wallet_kit.go:242
        	            				/Users/mohamedawnallah/Desktop/lnd/itest/lnd_onchain_test.go:285
        	            				/Users/mohamedawnallah/Desktop/lnd/itest/lnd_onchain_test.go:219
        	            				/Users/mohamedawnallah/Desktop/lnd/lntest/harness.go:353
        	            				/Users/mohamedawnallah/Desktop/lnd/itest/lnd_test.go:139
        	Error:      	Should be empty, but was [23a209dd3f537f457db65e06347c47c596f5e9a8e8a7a0b936c2d0475cc7945c 5339f96f1e6b441bfaa2664a96ca8cd4866ecd100e5162f864662543cbbce526]
        	Test:       	TestLightningNetworkDaemon/tranche00/01-of-1/btcd/cpfp
        	Messages:   	mempool not cleaned, please mine blocks to clean them all.
=== NAME  TestLightningNetworkDaemon
    lnd_test.go:147: Failure time: 2024-03-11 12:04:01.404
    lnd_test.go:155: =========> tests finished for tranche: 0, tested 1 cases, end height: 441
--- FAIL: TestLightningNetworkDaemon (49.35s)
    --- FAIL: TestLightningNetworkDaemon/tranche00/01-of-1/btcd/cpfp (36.50s)
FAIL
make: *** [itest-only] Error 255
```

ellemouton commented on itest/lnd_onchain_test.go:
```diff
@@ -278,9 +279,19 @@ func runCPFP(ht *lntest.HarnessTest, alice, bob *node.HarnessNode) {
-	bob.RPC.BumpFee(bumpFeeReq)
+	bumpFeeResp := bob.RPC.BumpFee(bumpFeeReq)
+
+	// The sweep tx hex on success should be set and be valid.
+	require.NotEmpty(ht, bumpFeeResp.SweepTxHex)
+	rawTxBytes, err := hex.DecodeString(bumpFeeResp.SweepTxHex)
+	require.NoError(ht, err, "bumpfee sweepTxHex invalid hex")
+	rawTx := &wire.MsgTx{}
+	err = rawTx.Deserialize(bytes.NewReader(rawTxBytes))
+	require.NoError(ht, err, "bumpfee sweepTxHex invalid")
+	require.Equal(ht, bumpFeeResp.SweepTxHex, rawTx.TxHash().String())
```
the reason this itest doesnt pass is cause it now waits for the sweep tx to be created which as you mentioned in the docs takes at least 30 seconds. But the itest `BumpFee` call times out after 30 seconds.

ellemouton replied:
nit: also, pls can you space things out a bit more for readability?

```
        // The sweep tx hex on success should be set and be valid.
	require.NotEmpty(ht, bumpFeeResp.SweepTxHex)

	rawTxBytes, err := hex.DecodeString(bumpFeeResp.SweepTxHex)
	require.NoError(ht, err, "bumpfee sweepTxHex invalid hex")

	rawTx := &wire.MsgTx{}
	err = rawTx.Deserialize(bytes.NewReader(rawTxBytes))
	require.NoError(ht, err, "bumpfee sweepTxHex invalid")
	require.Equal(ht, bumpFeeResp.SweepTxHex, rawTx.TxHash().String())
```

mohamedawnallah replied:
Thanks! Also Gonna space things out in [`itest/lnd_channel_force_close_test`](https://github.com/lightningnetwork/lnd/blob/5cb4811e86ed1607c1c9671756d7276256e2ec44/itest/lnd_channel_force_close_test.go#L149-L162) since it is related

ellemouton commented:
> rpc error: code = DeadlineExceeded desc = context deadline exceeded

Note that this was the hint to why the itest fails :)

mohamedawnallah commented on lnrpc/rpc_utils.go:
```diff
@@ -234,3 +236,17 @@ func CalculateFeeRate(satPerByte, satPerVByte uint64, targetConf uint32,
+// SerializeAndHexEncodeTx serializes the given wire transaction and returns
+// its hexadecimal representation. If serialization fails, an error is
+// returned.
+func SerializeAndHexEncodeTx(tx wire.MsgTx) (string, error) {
```
Okay. Initially, I considered this too. However, I opted to pass the copy because `SerializeAndHexEncodeTx` doesn't mutate the `tx` in place. I'm fine using the pointer to `wire.MsgTx` if it's commonly used in the codebase.

mohamedawnallah commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -227,6 +227,13 @@ var bumpFeeCommand = cli.Command{
+			Usage: "include the raw transaction hex for the " +
+				"sweep transaction on success. It waits for " +
+				"the batcher duration at least " +
+				"sweeper.batchwindowduration, default: 30s",
```
Thanks! Updated

mohamedawnallah commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1052,12 +1069,29 @@ func (w *WalletKit) BumpFee(ctx context.Context,
+			)
+			if err != nil {
+				return nil, err
+			}
```
Thanks! Updated

mohamedawnallah commented:
Thank you, @ellemouton, for your feedback. I've implemented the changes. However, I'm missing some understanding regards the bumpfee feature. It seems that the response from `lncli bumpfee` where `include_raw_tx` is true is not provided until we mine at least one block? 

To reproduce this, I ran the following command [including the unconfirmed funding transaction id]:

```shell
lncli wallet bumpfee --sat_per_vbyte 10 --include_raw_tx true e6a074b15cc28e1f37197f599fdd2021052c382a0b0307fa395ea12c291492a4:1
```

The command waits until the configured value `sweeper.batchwindowduration` (which is set to `10s` in my case) is reached. After that, the batcher batches the unconfirmed sweep transaction:

```shell
2024-03-12 12:10:28.925 [INF] SWPR: Candidate sweep set of size=1 (+0 wallet inputs), has yield=4.99960651 BTC, weight=445
2024-03-12 12:10:28.941 [INF] SWPR: Creating sweep transaction 42ce8661ee60ceaca54a6588393af0c4f7956788de8873a9e83fc4c55875a4f0 for 1 input (e6a074b15cc28e1f37197f599fdd2021052c382a0b0307fa395ea12c291492a4:1) using 2500 sat/kw, tx_weight=445, tx_fee=0.00001112 BTC, parents_count=0, parents_fee=0 BTC, parents_weight=0
2024-03-12 12:10:28.953 [INF] LNWL: Inserting unconfirmed transaction 42ce8661ee60ceaca54a6588393af0c4f7956788de8873a9e83fc4c55875a4f0
2024-03-12 12:10:28.958 [WRN] BTWL: Transaction 42ce8661ee60ceaca54a6588393af0c4f7956788de8873a9e83fc4c55875a4f0 not accepted by mempool: txn-already-in-mempool
2024-03-12 12:10:28.978 [INF] NTFN: New confirmation subscription: conf_id=3, txid=42ce8661ee60ceaca54a6588393af0c4f7956788de8873a9e83fc4c55875a4f0, num_confs=6 height_hint=501
```

However, it's still waiting without a response until the number of target confirmations is reached. In my case, I generated one block using bitcoin-cli:

```shell
bitcoin-cli -generate 1
```

After confirming the transaction, the logs of lnd show:

```shell
2024-03-12 12:11:55.865 [INF] LNWL: Marking unconfirmed transaction e6a074b15cc28e1f37197f599fdd2021052c382a0b0307fa395ea12c291492a4 mined in block 502
2024-03-12 12:11:55.866 [INF] CRTR: Pruning channel graph using block 00459e8025ac87baaf1b004d424c61afdd4f412f8c6a8999fa325ae1b85eb633 (height=502)
2024-03-12 12:11:55.878 [INF] CRTR: Block 00459e8025ac87baaf1b004d424c61afdd4f412f8c6a8999fa325ae1b85eb633 (height=502) closed 0 channels
2024-03-12 12:11:55.881 [INF] LNWL: Marking unconfirmed transaction 42ce8661ee60ceaca54a6588393af0c4f7956788de8873a9e83fc4c55875a4f0 mined in block 502
2024-03-12 12:11:56.111 [INF] NTFN: New block: height=502, sha=00459e8025ac87baaf1b004d424c61afdd4f412f8c6a8999fa325ae1b85eb633
2024-03-12 12:11:56.111 [INF] NTFN: Dispatching confirmed spend notification for outpoint=e6a074b15cc28e1f37197f599fdd2021052c382a0b0307fa395ea12c291492a4:1, script=1 0000000000000000000000000000000000000000000000000000000000000000 at current height=502: 42ce8661ee60ceaca54a6588393af0c4f7956788de8873a9e83fc4c55875a4f0[0] spending e6a074b15cc28e1f37197f599fdd2021052c382a0b0307fa395ea12c291492a4:1 at height=502
2024-03-12 12:11:56.112 [INF] NTFN: Canceling spend notification: spend_id=2, outpoint=e6a074b15cc28e1f37197f599fdd2021052c382a0b0307fa395ea12c291492a4:1, script=1 0000000000000000000000000000000000000000000000000000000000000000
2024-03-12 12:11:56.112 [INF] UTXN: Attempting to graduate height=502: num_kids=0, num_babies=0
```

Finally, I receive the following response from the bumpfee command:

```json
{
    "status":  "Successfully registered cpfp-tx with the sweeper",
    "sweep_tx_hex":  "02000000000101a49214292ca15e39fa07030b2a382c052120dd9f597f19371f8ec25cb174a0e6010000000000000000014bcbcc1d000000002251204a1c3a80076fca9bc75262f047ea86c54b84dc9fb5d07dca91b3d2cb0a4ed6df0140f6925b4abb71553d4d0b612522a006244a5976fc97e0068deec5de7321de152a2ac096ae82ed3b60419e2fa2472194868f90e7f5aa69832f947900f52af525c6f5010000"
}
```

Based on my assumption here increasing the default timeout here [`lntest/wait/timeouts_darwin.go` ](https://github.com/lightningnetwork/lnd/blob/master/lntest/wait/timeouts_darwin.go) wouldn't make the tests passed ?

mohamedawnallah commented:
Got it @ellemouton will do further investigation ask follow-up questions if any :)

guggero commented on rpcserver.go:
```diff
@@ -3874,14 +3874,12 @@ func (r *rpcServer) fetchWaitingCloseChannels(
 			if includeRawTx {
-				var txBuf bytes.Buffer
-				err = closingTx.Serialize(&txBuf)
+				closingTxHex, err = lnrpc.SerializeAndHexEncodeTx(
```
The linter complains about the line being too long. Maybe the function should instead be called `SerializeTx` and return bytes, then the hex encoding can be done here still (then it would be a more useful function since in most places we actually send raw bytes and not hex encoded transactions). I would've suggested to do the same here, but it looks like the other RPC that includes the raw transaction also uses a `string` instead of `bytes`. So I guess consistency is slightly better than wire-level efficiency in this case.

guggero reviewed (commented):
Looks good, once you find out what the issue with the failing itests is :+1:

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

ellemouton commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -973,11 +973,28 @@ func (w *WalletKit) BumpFee(ctx context.Context,
+				if err != nil {
+					return nil, err
+				}
+			case <-ctx.Done():
```
nit: new line between cases 🙏

ellemouton commented on itest/lnd_onchain_test.go:
```diff
@@ -278,9 +278,14 @@ func runCPFP(ht *lntest.HarnessTest, alice, bob *node.HarnessNode) {
 	}
-	bob.RPC.BumpFee(bumpFeeReq)
+
+	go func() {
+		bumpFeeResp := bob.RPC.BumpFee(bumpFeeReq)
+		require.NotEmpty(ht, bumpFeeResp.SweepTxHex)
+	}()
```
what is the reason that we need to spin this off in a goroutine? is it because BumpFee blocks until the tx is included in a block (meaning we need the "mine blocks" command below to run)? 

I think it is important that we know exactly why this blocks. Also - if it is blocking until the tx is mined, we should determine if this is in fact the behaviour that we want here. Is it possible to get the channel to return the raw hex at the time of tx construction rather? or do we wait until a block confirmation incase we update the structure of the tx? 

Regarding the goroutine - if we do decide that we in fact need to use a goroutine here, we should never do it without tracking the go routine - as it currently stands, we have no idea if the code in here ever even gets run or that it completes. So you would need to add a wait group or something so that we can wait on it later.

mohamedawnallah replied:
> What is the reason that we need to spin this off in a goroutine? is it because BumpFee blocks until the tx is included in a block (meaning we need the "mine blocks" command below to run)?

Yes, that was my intent!

mohamedawnallah replied:
> Regarding the goroutine - if we do decide that we in fact need to use a goroutine here, we should never do it without tracking the go routine - as it currently stands, we have no idea if the code in here ever even gets run or that it completes. So you would need to add a wait group or something so that we can wait on it later.

Alright 👍

mohamedawnallah replied:
> Is it possible to get the channel to return the raw hex at the time of tx construction rather?

Since the `resultChan` is not accessible in [sweep/sweeper.go/createSweepTx](https://github.com/lightningnetwork/lnd/blob/5599b3c9e2fa9dade884e1a6de96479c9697e35a/sweep/sweeper.go#L1166-L1169) where we create sweep transaction may be not feasible to get the raw hex at the time of tx construction ?

mohamedawnallah replied:
> do we wait until a block confirmation incase we update the structure of the tx?

 I don't see any mutation for the structure of spent sweep transaction after the transaction confirmation (monitored by `monitorSpend` function) in [handleInputSpent](https://github.com/lightningnetwork/lnd/blob/5599b3c9e2fa9dade884e1a6de96479c9697e35a/sweep/sweeper.go#L1572).

ellemouton reviewed (changes requested):
See my comment regarding the goroutine in the itest 😊

guggero commented:
Removed my request for review for now. Please re-request when the PR is ready.

mohamedawnallah commented:
Thanks again @ellemouton for your feedback! After deeper investigation (summarized into the following diagram) we need to wait for a message sent to `resultChan` till the given sweep input is confirmed in order to get the transaction id.

![bumpfee drawio](https://github.com/lightningnetwork/lnd/assets/69568555/fdfec029-ab3f-4743-a333-a601c351043b)

mohamedawnallah commented:
@ellemouton What should we do in this case? Should we go with using go routines + wait groups when we test CPFP?

ellemouton commented:
@mohamedawnallah  - great investigation!! 👏 

So I think the main thing here is that we definitely _dont_ want an RPC call to ever block while waiting for a block confirmation. From your investigation, it seems that this is what will now happen. This seems to be because it waits until the tx is in its final state (ie, has been confirmed & therefor we wont change it again (add more inputs/outputs etc) before sending the hex back. 

So perhaps this PR will need to add a bit more so that we can just get the first tx hex as soon as it is available. If it ends up changing due to being RBF'd, that's fine. I think the users just want an initial "ok something did happen and here is the hex of the first tx it is trying" type of thing. 

Does that make sense?

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

ziggie1984 commented:
Thanks for the PR great work so far and nice analysis (with which tool did you create the flow-chart ?) 💪,

Having the new PR-Sweeper Series in mind (which will land in 18), this waiting period will change. However I am not sure if we really wanna return the Sweep-Transaction here, because it's not constant, it might change over the course it is in the mempool. So I rather think of something were we depict the current sweep-transaction in the `wallet pendingsweeps` response. There we list all the inputs and I think there is the right place to also show the raw-transaction to a corresponding input sweep. Wdyt @yyforyongyu ?

mohamedawnallah commented:
> which tool did you create the flow-chart ?

I used [draw.io](https://www.drawio.com/) to create the flow chart.

ellemouton commented:
>  So I rather think of something were we depict the current sweep-transaction in the wallet pendingsweeps response. 

I think that's a great idea

mohamedawnallah commented:
**Mapping Sweep Input to New Sweeper Transaction for Hex**

How do we map the sweep input (outpoint or reference to the previous transaction output that we want to bump its fee) from `wallet pendingsweeps` to the newly created sweeper transaction to obtain the hex?

**Further Details**

Let's ensure we're on the same page. My understanding of bumping fees uses the term "outpoint" as a synonym for "sweep input" (please correct me if I'm wrong).

![Untitled-2024-04-19-0518](https://github.com/lightningnetwork/lnd/assets/69568555/4ed151a3-e4a0-4fdf-a93d-dd40e3fb6c6a)

Our goal, when running the bump fee command, is to return the new sweep transaction hex so the user can rebroadcast it if it fails to go through. To obtain the new sweep transaction hex, we need to identify the outpoint or sweep input from `lncli wallet pendingsweeps`.

Currently, we don't have access to the new sweep transaction. How can we access it to obtain the raw hex?

I've explored `lncli listchaintxns`, where sweep transactions are ending with the label `*:sweep`. We can filter these, but there might be multiple transactions referencing the same previous outpoint (e.g., CPFP to RBFP). How should we index the most recent, intended transaction when bumping the fee?

![Untitled-2024-04-19-0518](https://github.com/lightningnetwork/lnd/assets/69568555/bc675fe0-9cff-4039-af35-5ce2198bde02)

ziggie1984 commented:
Let's put this PR on Hold for a bit, we want to redesign the sweeper RPC cmds in 18.1, because things changed quite a bit in LND 18. But the main idea is, that bumping the fee of an input will not report back immediately the new hex transaction, rather querying the sweeper stats should reveal all the infos for a specific sweep input (e.g. current sweep hex), so that the user can cpy&paste it from there.

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready
